## Exercise1

Suppose we have a neural network for predicting the time between hospital readmissions for a certain subset of patients. Such a network would depends on many parameters, but for simplicity, let's assume there's only two parameters: a and b. Each parameter lies between 0 and 1 and represents the weights of two "synapses" in our model.

Using the API at e.g.

http://ramcdougal.com/cgi-bin/error_function.py?a=0.4&b=0.2 (Links to an external site.)

one can find the prediction errors in arbitrary units of running our model with specified parameters (here a=0.4 and b=0.2, but you can change that).

Implement a two-dimensional version of the gradient descent algorithm to find optimal choices of a and b. (7 points) (Optimal here means values that minimize the error.) See slides 12 for a 1-dimensional implementation of the algorithm. Explain how you estimate the gradient given that you cannot directly compute the derivative (3 points), identify any numerical choices -- including but not limited to stopping criteria -- you made (3 points), and justify why you think they were reasonable choices (3 points).

It so happens that this error function has a local minimum and a global minimum. Find both locations (i.e. a, b values) querying the API as needed (5 points) and identify which corresponds to which (2 point). Briefly discuss how you would have tested for local vs global minima if you had not known how many minima there were. (2 points)

Important: a key part of this problem is the implementation of the algorithm; do not use a Python library for gradient descent or copy gradient descent code from a website.

(Disclaimer: the API function is not actually testing a neural network.)

(Hint: remember the gradient is just a vector of 1D derivatives.)

Note: Please be kind to my server. Do not systematically sweep through the parameter space. Using gradient descent will only require a moderate number of queries.

Note: Explicitly specify a User-Agent header or you'll get a 406 error; i.e. do something like:

## Response

In [9]:
## Importing relevant libraries 

import requests
import numpy as np
import pandas as pd
from numpy.linalg import norm

#### Using the APIs to find the prediction errors in arbitrary units

In [10]:
def f(a,b):

    parameters = {
        'a':a,
        'b':b
    }
    headers = {
        "User-Agent": "MyScript"
    }
    url = "http://ramcdougal.com/cgi-bin/error_function.py"
    result = requests.get(url,params = parameters,headers=headers)
    return float(result.text)

In [11]:
f(0.4, 0.2)

1.294915

#### Function for derivative of f(a,b) with respect to a

In [12]:
def derivativeA(a, b, h):
    return (f(a + h,b) - f(a,b))/h

#### Function for derivative of f(a,b) with respect to b

In [13]:
def DerivativeB(a, b, h):
    return (f(a,b + h) - f(a,b))/h

#### 2D Vector containing the derivative of both A and B

In [14]:
def finalderivative(a, b, h):
    return np.array([derivativeA(a, b, h), DerivativeB(a, b, h)]) 

#### Two-dimensional version of the gradient descent algorithm to find optimal choices of a and b.

In [15]:
def gradient_descent(init_step, alpha, h = 1e-4, tolerance = 1e-8):
    previous_step_size = init_step - 10 * tolerance
    step = init_step
     
    #values for the gradient descent algorithm 
    step_a = []
    step_b = []
    end_result = []
    n_iter = 0
    max_iter = 1000
    while norm(step - previous_step_size) > tolerance and n_iter < max_iter :
        previous_step_size = step
        step = step - alpha * finalderivative(step[0], step[1], h)
        step_a.append(step[0])
        step_b.append(step[1])
        end_result.append(f(step[0], step[1]))
        n_iter+=1
        
    print (f"The minimum value {end_result[-1]} occurs at :")
    print (f"a = {step_a[-1]}, b = {step_b[-1]}.")

#### Examples

In [16]:
gradient_descent(np.array([0.001, 0.999]), 0.1)

The minimum value 1.100000005 occurs at :
a = 0.21595000000038012, b = 0.6889500399996086.


In [17]:
gradient_descent(np.array([0.999, 0.001]), 0.1)

The minimum value 1.000000015 occurs at :
a = 0.7119500099997124, b = 0.1689500000000278.


In [18]:
gradient_descent(np.array([0.09, 0.05]), 0.1)

The minimum value 1.100000005 occurs at :
a = 0.21595000000015233, b = 0.6889499600005788.


In [19]:
gradient_descent(np.array([0.05, 0.09]), 0.1)

The minimum value 1.100000005 occurs at :
a = 0.2159499999995262, b = 0.6889499599998726.


#### Gradient Descent is an optimization algorithm to find the minimum of a function. We start with a random point on the function and move in the negative direction of the gradient of the function to reach the local/global minima.

#### Qa. Explain how you estimate the gradient given that you cannot directly compute the derivative? 
Ans.  I used the definition of derivative f'(x) = f(x + h) - f(x)/h and partial derivative function to estimate the gradient function. The parameter value (h) was set to be small. 

Reference: https://www.khanacademy.org/math/multivariable-calculus/multivariable-derivatives/partial-derivative-and-gradient-articles/a/the-gradient and the cheatsheet provided. 


#### Qb. Identify any numerical choices -- including but not limited to stopping criteria -- you made (3 points), and justify why you think they were reasonable choices (3 points).

Ans. Learning Rate: It determines how fast or slow we will move towards the optimal weights. If the learning rate is very large we will skip the optimal solution. If it is too small we will need too many iterations to converge to the best values.The obvious way to find a desirable or optimal learning rate is through trial and error. For the gradient descent the default learning rate is 0.01 or 0.1. Therefore, I have used 0.01 as the alpha. 

Stopping Criteria: I have set the max iteration to 1000. It makes the algorithm to stop iterating and return the result before max_iter is reached if the vector update in the current iteration is less than or equal to tolerance. 


#### Qc. Identify which corresponds to which (2 point). Briefly discuss how you would have tested for local vs global minima if you had not known how many minima there were. (2 points)

Ans. In a gradient descent model the smallest value is global minimum while other points are local minimum. Trying different combinations of the value while keeping the other parameters same would help in identifying the local vs global minima. Aso, gradient descent is able to find local minima most of the time and not global minima because the gradient does not point in the direction of the steepest descent. Current techniques to find global minima either require extremely high iteration counts or a large number of random restarts for good performance.The global minima is 1.000000015 which occurs at : a = 0.7119500099997124, b = 0.1689500000000278 while the local minima is 1.100000005 which occurs at : a = 0.21595000000038012, b = 0.6889500399996086 or at a = 0.2159499999995262, b = 0.6889499599998726.